### PR TITLE
 Make CrateNum allocation more thread-safe. 

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR makes sure that we can't have race conditions when assigning CrateNums. It's a slight improvement but a larger refactoring of the CrateStore/CrateLoader infrastructure would be good, I think.

r? @Zoxc

### CODE REVIEW

**Comment by Zoxc** [comment]
I would use the length of the `CStore.metas` field as the next `CrateNum` instead of using a separate counter. `alloc_new_crate_num` would then push `None` to `metas` before returning and `set_crate_data` should be changed to no longer expand `metas`.

**Comment by michaelwoerister** [comment]
Good idea, I'll do that.

**Comment by michaelwoerister** [comment]
Updated with the new approach. Much nicer this way.

File: src/librustc_metadata/cstore.rs [hunk 1]
```diff
@@ -96,32 +96,30 @@ pub struct CStore {
 impl CStore {
     pub fn new(metadata_loader: Box<MetadataLoader + Sync>) -> CStore {
         CStore {
-            metas: RwLock::new(IndexVec::new()),
+            metas: RwLock::new(IndexVec::from_elem_n(None, 1)),
```
Comment (Zoxc) [inline src/librustc_metadata/cstore.rs]: The first entry is for `LOCAL_CRATE` here right? Do we call `set_crate_data` for that?

Comment (michaelwoerister) [reply src/librustc_metadata/cstore.rs]: Yes, that's for the local crate. The entry will always be `None`. It's just there to make array indices match `CrateNum`s.

Comment (Zoxc) [reply src/librustc_metadata/cstore.rs]: Could you add that as a comment?

Comment (michaelwoerister) [reply src/librustc_metadata/cstore.rs]: Done.

**Comment by michaelwoerister** [comment]
Tidy fixed.

@bors r=Zoxc

**Pull request merged**